### DISCUSSION TITLE
Get output from a running jest process under Docker

### DISCUSSION THREAD

**Discussion by navalex** [orta/vscode-jest#1123] [Q&A]
Hello,

My whole project is running under several Docker Compose containers for development. I made a dedicated container that run a "jest --watchAll --collectCoverage".
I can't run my jest outside it's container because my node_modules are not made to work on my host machine (which is the most of time a macbook).

So my question is: is there a way to collect the outputs of my jest containers to make the plugin show me the coverage and test outputs ? Or maybe a way to run the plugin in the container ? I don't really know how does it works so hard to imagine a solution.

Thanks

EDIT:
For the moment I just moved to Coverage-Gutter, but it would be awesome to have my test feedback inside vscode too:)

**Comment by connectdotz** [comment]
I have never done it myself, but I know some of our users did. Here are the instructions ChatGPT generated, you could give it a try and let me know if it works:

---
To run the `vscode-jest` extension within a Docker container as part of your development environment, leveraging Visual Studio Code's **Remote - Containers** extension, you'll follow a setup that ensures `vscode-jest` operates smoothly with your Jest tests running inside the container. Here's a step-by-step guide:

### Prerequisites

- Ensure you have Docker and Docker Compose installed on your host machine.
- Install Visual Studio Code (VSCode) on your host machine.
- Install the **Remote - Containers** extension in VSCode.

### 1. Configure Your Docker Environment

1. **Dockerfile**: Make sure your Dockerfile sets up a Node.js environment, installs dependencies, and configures your project to run Jest tests. You might already have this from setting up your Dockerized environment.

2. **docker-compose.yml**: Ensure your `docker-compose.yml` file is set up to build your service(s), including your application and any other dependent services (like databases).

### 2. Setup `devcontainer.json`

1. **Create a `.devcontainer` Folder**: Inside your project's root directory, create a folder named `.devcontainer`.

2. **Add `devcontainer.json`**: Inside the `.devcontainer` folder, create a file named `devcontainer.json`. This file will configure your VSCode environment when you open your project in a container.

3. **Configure `devcontainer.json`**: Add the necessary configurations to specify your Docker setup and to install the `vscode-jest` extension inside your container. Here's an example configuration:

```json
{
    "name": "Your Project Name",
    "dockerComposeFile": "../docker-compose.yml",
    "service": "your-service-name",
    "workspaceFolder": "/path/to/your/workspace/in/container",
    "extensions": [
        "Orta.vscode-jest"
    ],
    "settings": {
        // Optional: Any VSCode settings you want to apply in the container
    },
    "postCreateCommand": "npm install",
    "shutdownAction": "stopCompose"
}
```

Adjust the `dockerComposeFile`, `service`, `workspaceFolder`, `jest.pathToJest`, `jest.pathToConfig`, and `jest.rootPath` to match your project's setup. The `postCreateCommand` ensures dependencies are installed when the container is created.

### 3. Start Your Project in a Container

1. **Open Command Palette**: With your project open in VSCode, press `F1` or `Ctrl+Shift+P` (`Cmd+Shift+P` on Mac) to open the Command Palette.

2. **Reopen in Container**: Type "Remote-Containers: Reopen in Container" and press enter. VSCode will start the process of building your container (if necessary) and then open your project inside the container.

3. **Verify `vscode-jest` is Running**: Once your project is open in the container, the `vscode-jest` extension should be active. You'll see Jest's status in the VSCode status bar, and it should automatically run your tests. Ensure your Jest configuration and tests are set up to run in this environment.

### 4. Working with Jest in the Container

- The `vscode-jest` extension will monitor your test files for changes and run tests automatically.
- Use VSCode's integrated terminal if you need to run Jest or other commands manually. It will execute within the context of your container.

### Troubleshooting

- **Jest Configuration**: Ensure your `jest.config.js` is correctly set up for a containerized environment. Paths and any environment-specific configurations may need adjustments.
- **Dependencies**: If you encounter issues related to dependencies not being found, check that your container's `npm install` or `yarn install` ran successfully during the container's setup phase.

By following these steps, you'll have the `vscode-jest` extension running within your Docker container, providing an integrated experience for running and debugging Jest tests directly in VSCode, leveraging the consistency and isolation benefits of containerized development environments.

---

**Reply by navalex** [reply]
I give it a try, keep you in touch

EDIT:

I will probably work but it's not what i'm looking for because I don't want to open my vscode inside a container. But i managed to fix half of my issue by using Gutter-Coverage, which use my coverage output folder to keep sync.

**Reply by connectdotz** [reply]
my impression is that the instruction above is for running vscode outside of the container, while the extension itself will run within the container. The vscode-jest extension runs jest by spawning a node child process with your jest command line, so it will need to run within the container.
